### ISSUE TITLE
Using FsCheck.Xunit in C#, how do I clean up/tear down after a test has been executed?

### ISSUE THREAD

**Issue by CAP-3Shape** [fscheck/FsCheck#514]
I am using FsCheck - specifically the FsCheck.Xunit package - in my C# code. I want to run some setup before my tests execute and some cleanup/teardown code after my tests have all run.

NUnit has a [CleanUp] attribute; Xunit uses _Dispose_ to clean up. In some frameworks I think it's called [TearDown]. Is there anything of this sort in FsCheck.Xunit?

When I rewrote one Xunit test class to use FsCheck, I could see that my _Dispose_ method no longer got executed.

In the [documentation ](https://fscheck.github.io/FsCheck/reference/index.html) there is something called "TearDown", but it is completely opaque to me how it is supposed to be used.

I [posted this on Stack Overflow](https://stackoverflow.com/questions/60702341/using-fscheck-xunit-in-c-how-do-i-clean-up-tear-down-after-all-tests-have-been) and Kurt Schelfthout asked me to report it as an issue here. 

Is it a bug in FsCheck.Xunit? Ought it to execute my _Dispose_ method like a regular Xunit test?

I am using: 

- FsCheck.Xunit 2.14.1
- .NET core 3.1
- C# 8.0
- Visual Studio 2019 version 16.4.6

**Comment by kurtschelfthout** [comment]
Yes, it looks like after replacing FactAttribute with Property, xunit no longer calls Dispose. 

I’ll have to take a look why. 

Meanwhile you may be able to work around by not using FsCheck.Xunit at all, by defining properties inside a normal Xunit Fact method, using eg QuickCheckThrowOnFailure and the Prop API.

**Comment by ploeh** [comment]
Wouldn't that cause the property to run (by default) 100 times, and then do a single cleanup?

**Comment by CAP-3Shape** [comment]
> Wouldn't that cause the property to run (by default) 100 times, and then do a single cleanup?

It would be great to have a way to clean up after each run AND a way to clean up after all runs, similar to the [TestCleanup] and [ClassCleanup] that some test frameworks have. I don't know how feasible that is in Xunit.

**Comment by ploeh** [comment]
IIRC xUnit.net has nothing like `[ClassCleanup]`...

**Comment by kurtschelfthout** [comment]
> Wouldn't that cause the property to run (by default) 100 times, and then do a single cleanup?
Indeed. I thought that was the original question.

If you want to setup/cleanup before/after each individual FsCheck test, then doing it manualy is your best bet. e.g.

```csharp
[Property]
public void MyTest(a: int, b:int) {
    // setup
    try {
        // assert something
    }
    finally {
        // cleanup
    }
}
```

Regardless that Dispose is no longer called seems like a bug :)

**Comment by kurtschelfthout** [comment]
Related - I just remembered that for testing mutable objects with FsCheck using the model based testing approach, you can tear down each created object.

With model-based testing, FsCheck in each test generates an initial object (typically by calling a ctor),  applies a random sequence of operations, and then optionally calls a `TearDown` method. The latter would typically call `Dispose`.

https://fscheck.github.io/FsCheck/StatefulTestingNew.html
https://fscheck.github.io/FsCheck/reference/fscheck-experimental-machine-2.html

**Comment by kurtschelfthout** [comment]
The fix in #519 has been released in [2.14.3](https://www.nuget.org/packages/FsCheck.Xunit/2.14.3)
